### PR TITLE
Define a health endpoint for hook and tot services

### PR DESCRIPTION
**Pull request by ccojocar**

### CODE REVIEW

File: prow/cmd/tot/main.go [hunk 1]
```diff
@@ -301,6 +301,7 @@ func main() {
 	}
 
 	http.HandleFunc("/vend/", s.handle)
+	http.HandleFunc("/healthz/ready", func(w http.ResponseWriter, r *http.Request) { fmt.Fprint(w, "OK") })
```
Comment (stevekuznetsov) [inline prow/cmd/tot/main.go]: Can we please follow the example in `deck` and support both readiness and health endpoints?

https://github.com/kubernetes/test-infra/blob/1f4a1eea2d8aee8b723e274a79c1cb722d091023/prow/cmd/deck/main.go#L161-L163

Comment (stevekuznetsov) [reply prow/cmd/tot/main.go]: (this is an opportunity to DRY it out as well a la the pprof endpoint)

**Comment by fejta** [comment]
/ok-to-test

Some have suggested the idea of deleting tot to simplify the system. How important is this component to you?

File: prow/cmd/tot/main.go (hunk 1 above)
Comment (ccojocar) [reply prow/cmd/tot/main.go]: I defined a helper for helath endpoints in `pjutil` package and integrated in deck and tot.

**Comment by ccojocar** [comment]
>Some have suggested the idea of deleting tot to simplify the system. How important is this component to you?

It is not very import. I wanted it to get it running with Jenkins X prow helm chart just to validate that it is properly vending build IDs for pipeline controller.

**Comment by fejta** [comment]
You should be able to just drop it. The rest of prow functions fine without this service (we generate snowflake IDs from a monotonically increasing pure function), and this is how prow.k8s.io runs now.

LGTM but letting Steve do the rest of the review

File: prow/pjutil/health.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+)
+
+// Health keeps a request multiplexer for health liveness and readiness endpoints
+type Health struct {
+	healthMux *http.ServeMux
+}
+
+// NewHealth creates a new health request multiplexer and starts serving the liveness endpoint
+// on the given port
+func NewHealth(port int) *Health {
```
Comment (stevekuznetsov) [inline prow/pjutil/health.go]: Why do we need to parameterize this?

**Comment by stevekuznetsov** [comment]
>  Jenkins X prow helm chart 

@ccojocar is that something you want to see checked in to `test-infra`? Is it different from our `starter.yaml`?

File: prow/pjutil/health.go (hunk 2 above)
Comment (ccojocar) [reply prow/pjutil/health.go]: It might be that some consumers of this health endpoint might want to configure a different port than 8081.

**Comment by ccojocar** [comment]
> is that something you want to see checked in to test-infra? Is it different from our starter.yaml?

It seems that `tot` is not defined there. Having liveness and readiness probes for it, it is useful.

File: prow/pjutil/health.go (hunk 2 above)
Comment (fejta) [reply prow/pjutil/health.go]: a) if every instance of this uses 8081 it suggests we don't actually need this argument right now
b) at least use a constant?
c) only tot and deck have health checks?

Comment (ccojocar) [reply prow/pjutil/health.go]: I exacted the port into a constant.

Comment (ccojocar) [reply prow/pjutil/health.go]: c) also `hook` has a health check but on the root path

I integrated this health check also into the hook service, but still keep the old one for compatibility. 

Also added the liveness and readiness probes to hook deployment.

**Comment by stevekuznetsov** [comment]
LGTM -- we need to merge the code and get it deployed before we change the YAML or the deployments will fail

**Comment by ccojocar** [comment]
Do you want me to send the YAML in a follow up PR?

**Comment by fejta** [comment]
The yaml PR needs to happen both a) after this PR and b) after we deploy a prow release that has this new functionality. AKA tomorrow

**Comment by ccojocar** [comment]
OK. Let me separate it in a different PR.

**Comment by ccojocar** [comment]
@fejta @stevekuznetsov I moved the deployment in this PR https://github.com/kubernetes/test-infra/pull/12069.

**Comment by fejta** [comment]
/hold
Feel free Steve/Cosmin to remove the hold when you're ready for it to merge (and/or provide more comments)

**Pull request merged**